### PR TITLE
Kernel Multiprocess (Part 1) - Persist memory & core timing

### PR DESCRIPTION
**Pull request by bunnei**
This is the first batch of changes in support of multi-process emulation.

### Background

Previously, a single "emulation session" (i.e., running a game) would consist of emulating Switch boot (initializing memory, kernel, services, GPU, etc.), creating a single new process (for the emulated title), and then running that process. Ending that session would consist of forcefully tearing all of that down at once, such that we are back at the initial state. Launching a subsequent title would repeat all of this.

We're planning to rearchitect emulation such that launching yuzu is much like booting Switch, and what was previously an emulation session is purely launching a new process for the chosen title. Furthermore, shutting down is closing just the running title process. The intent of these changes is more accurate full system emulation, such as by allowing us to emulate multiple processes at once. Furthermore, this should allow a better user-experience -- faster and more stable game launches, eventual support for system apps such as qlaunch, etc.

### Change summary

With this first set of changes, we focus on partially persisting system emulation across game boots. Now, when you launch yuzu, we will instantiate:
* Device memory (Switch's 4GB DRAM)
* Core timing

These will subsequently be re-used on subsequent game launches.

For these changes, we mostly needed to focus on improving our memory emulation. These changes include:
* Accurately tracking memory allocations, to ensure they are freed on process shutdown.
* Adding/refreshing various kernel memory modules to latest RE (KMemoryBlock, KMemoryBlockManager, KDynamicPageManager, KDynamicSlabHeap, KDyamicResourceManager, KPageTable).
* Reworking CoreTiming to allow multiple emulation sessions.
* Improved emulation of KThread shutdown & DPC.

Some collateral changes from these include:
* Some yuzu settings change the fundamental nature of full-system emulation (e.g. multicore vs. singlecore). For these settings changes, we introduce a new path to fully re-initialize the system on game launch if necessary. Hopefully this can go away one day.
* All new/changed code has been updated to use the newly introduced result macros.
* Various other code cleanups & improvements.

### Testing & expectations
Components of kernel emulation have been refreshed for accuracy (primarily around memory management), so it's possible some bugs are fixed. That being said, I do not expect these changes to have much of a noticeable impact on game compatibility. Instead, please focus on:
* Improved game launch performance & reliably, particularly across multiple game launches.
* Game shutdown performance & reliability. These changes will introduce some lag in game shutdown, as we are now doing more work in an attempt to emulate process shutdown (still a work in progress), rather than just terminating the full emulation state forcibly.
* Overall memory usage -- Expect it to be higher when yuzu is idle, but unchanged or lower during emulation.

### CODE REVIEW

File: src/core/hle/kernel/k_memory_block.h [hunk 1]
```diff
@@ -261,113 +297,349 @@ class KMemoryBlock final {
...
+        m_disable_merge_attribute = static_cast<KMemoryBlockDisableMergeAttribute>(
+            m_disable_merge_attribute & KMemoryBlockDisableMergeAttribute::AllRight);
     }
 
-    constexpr bool HasSameProperties(const KMemoryBlock& rhs) const {
-        return state == rhs.state && perm == rhs.perm && original_perm == rhs.original_perm &&
-               attribute == rhs.attribute && ipc_lock_count == rhs.ipc_lock_count &&
-               device_use_count == rhs.device_use_count;
+    constexpr void UpdateDeviceDisableMergeStateForShareLeft(
+        [[maybe_unused]] KMemoryPermission new_perm, bool left, [[maybe_unused]] bool right) {
```
Comment (Tachi107) [inline src/core/hle/kernel/k_memory_block.h:470]: If these parameters are always unused, you can simply do

```suggestion
        KMemoryPermission /*new_perm*/, bool left, bool /*right*/) {
```

**Review by liamwhite — APPROVED** [review]
Code looks great. Tested booting 20 things in a row, no issues besides a minor delay on shutdown (a spinner would be nice, but out of scope for this PR.)

File: src/core/hle/kernel/k_dynamic_page_manager.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            std::memset(GetPointer<PageBuffer>(m_address) + i, 0, PageSize);
+
+            // Set the bit for the free page.
+            m_page_bitmap.SetBit(i);
+        }
+
+        R_SUCCEED();
+    }
+
+    constexpr VAddr GetAddress() const {
```
Comment (lioncash) [inline src/core/hle/kernel/k_dynamic_page_manager.h:83]: Given the constructor isn't constexpr, none of these getters can ever be executed in a constexpr context.

File: src/core/hle/kernel/k_dynamic_page_manager.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+class KDynamicPageManager {
+public:
+    class PageBuffer {
+    private:
+        u8 m_buffer[PageSize];
+    };
+    static_assert(sizeof(PageBuffer) == PageSize);
+
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_dynamic_page_manager.h:23]: These should be after the public interface below to follow our convention.

File: src/core/hle/kernel/k_dynamic_resource_manager.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+template <typename T, bool ClearNode = false>
+class KDynamicResourceManager {
+    YUZU_NON_COPYABLE(KDynamicResourceManager);
+    YUZU_NON_MOVEABLE(KDynamicResourceManager);
+
+public:
+    using DynamicSlabType = KDynamicSlabHeap<T, ClearNode>;
+
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_dynamic_resource_manager.h:20]: Ditto

File: src/core/hle/kernel/k_dynamic_slab_heap.h [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+template <typename T, bool ClearNode = false>
+class KDynamicSlabHeap : protected impl::KSlabHeapImpl {
+    YUZU_NON_COPYABLE(KDynamicSlabHeap);
+    YUZU_NON_MOVEABLE(KDynamicSlabHeap);
+
+private:
+    using PageBuffer = KDynamicPageManager::PageBuffer;
+
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_dynamic_slab_heap.h:22]: Ditto

File: src/core/hle/kernel/k_memory_block_manager.h [hunk 6]
```diff
@@ -4,63 +4,154 @@
...
+
+    MemoryBlockTree m_memory_block_tree;
+    VAddr m_start_address{};
+    VAddr m_end_address{};
+};
 
-    [[maybe_unused]] const VAddr start_addr;
-    [[maybe_unused]] const VAddr end_addr;
+class KScopedMemoryBlockManagerAuditor {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_memory_block_manager.h:143]: Ditto

File: src/core/core.cpp [hunk 7]
```diff
@@ -133,6 +133,48 @@ struct System::Impl {
...
+        const auto current_time =
+            std::chrono::duration_cast<std::chrono::seconds>(posix_time).count();
+        Settings::values.custom_rtc_differential =
+            Settings::values.custom_rtc.value_or(current_time) - current_time;
+
+        // Create a default fs if one doesn't already exist.
+        if (virtual_filesystem == nullptr)
+            virtual_filesystem = std::make_shared<FileSys::RealVfsFilesystem>();
+        if (content_provider == nullptr)
+            content_provider = std::make_unique<FileSys::ContentProviderUnion>();
```
Comment (Kelebek1) [inline src/core/core.cpp:154]:
```suggestion
        if (virtual_filesystem == nullptr) {
            virtual_filesystem = std::make_shared<FileSys::RealVfsFilesystem>();
        }
        if (content_provider == nullptr) {
            content_provider = std::make_unique<FileSys::ContentProviderUnion>();
        }
```

File: src/core/hle/kernel/svc_common.h [hunk 8]
```diff
@@ -14,8 +14,13 @@ namespace Kernel::Svc {
 
 using namespace Common::Literals;
 
-constexpr s32 ArgumentHandleCountMax = 0x40;
-constexpr u32 HandleWaitMask{1u << 30};
+enum {
+    HandleWaitMask = (1u << 30),
```
Comment (lioncash) [inline src/core/hle/kernel/svc_common.h:18]: What does this do that leaving it as a `constexpr u32` can't do?

File: src/core/hle/kernel/svc_types.h [hunk 9]
```diff
@@ -95,6 +95,19 @@ constexpr inline s32 IdealCoreNoUpdate = -3;
 constexpr inline s32 LowestThreadPriority = 63;
 constexpr inline s32 HighestThreadPriority = 0;
 
+constexpr inline s32 SystemThreadPriorityHighest = 16;
+
+enum ProcessState : u32 {
```
Comment (lioncash) [inline src/core/hle/kernel/svc_types.h:100]: Shouldn't this be an enum class?

**Review by lioncash — COMMENTED** [review]
Looks good. Just a few minor nits

**Pull request merged**